ruby/ql/lib/codeql/ruby/security/CleartextLoggingCustomizations.qll

```diff
@@ -0,0 +1,311 @@
+    MaskingReplacerSanitizedNode() {
+      exists(MaskingReplacerSanitizer maskCall, Variable v |
+        maskCall.getMethodName() = ["sub!", "gsub!"] and
+        v = maskCall.getReceiver().asExpr().getExpr().(VariableReadAccess).getVariable() and
+        v = this.asExpr().getExpr().(VariableReadAccess).getVariable() and
+        maskCall.asExpr().getASuccessor*() = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * Gets the name of a  method that would be falsely marked as non-sensitive
+   * by `notSensitiveRegexp`.
+   */
+  private predicate nonSensitiveMethodNameExclusion(string name) { name = ["[]", "[]="] }
+
+  /**
+   * A call that might obfuscate a password, for example through hashing.
+   */
+  private class ObfuscatorCall extends Sanitizer, DataFlow::CallNode {
+    ObfuscatorCall() {
+      this.getMethodName().regexpMatch(notSensitiveRegexp()) and
+      not nonSensitiveMethodNameExclusion(this.getMethodName())
+    }
+  }
+
+  /**
+   * A data flow node that does not contain a clear-text password, according to its syntactic name.
+   */
+  private class NameGuidedNonCleartextPassword extends NonCleartextPassword {
+    NameGuidedNonCleartextPassword() {
+      exists(string name | name.regexpMatch(notSensitiveRegexp()) |
+        // accessing a non-sensitive variable
+        this.asExpr().getExpr().(VariableReadAccess).getVariable().getName() = name
+        or
+        // dereferencing a non-sensitive field
+        this.asExpr()
+            .getExpr()
+            .(ElementReference)
+            .getArgument(0)
+            .getConstantValue()
+            .getStringOrSymbol() = name
+        or
+        // calling a non-sensitive method
+        this.(DataFlow::CallNode).getMethodName() = name and
+        not nonSensitiveMethodNameExclusion(name)
+      )
+      or
+      // avoid i18n strings
+      this.asExpr()
+          .getExpr()
+          .(ElementReference)
+          .getReceiver()
+          .getConstantValue()
+          .getStringOrSymbol()
+          .regexpMatch("(?is).*(messages|strings).*")
+    }
+  }
+
+  /**
+   * A data flow node that receives flow that is not a clear-text password.
+   */
+  private class NonCleartextPasswordFlow extends NonCleartextPassword {
+    NonCleartextPasswordFlow() {
+      any(NonCleartextPassword other).(DataFlow::LocalSourceNode).flowsTo(this)
+    }
+  }
+
+  /**
+   * A data flow node that does not contain a clear-text password.
+   */
+  abstract private class NonCleartextPassword extends DataFlow::Node { }
+
+  // `writeNode` assigns pair with key `name` to `val`
+  private predicate hashKeyWrite(DataFlow::CallNode writeNode, string name, DataFlow::Node val) {
+    exists(SetterMethodCall setter |
+      setter = writeNode.asExpr().getExpr() and
+      // hash[name]
+      setter.getArgument(0).getConstantValue().getStringOrSymbol() = name and
+      // val
+      setter.getArgument(1).(Assignment).getRightOperand() = val.asExpr().getExpr()
+    )
```

Comment:
I think this will also include things like `foo.name = val`, but maybe that's fine?

Reply:
In practice this doesn't match against those assignments, though this is more down to the details of how we desugar these assignments than an intentional choice. Something like:
```
      // foo.name = val
      setter.getMethodName() = name + "=" and
      setter.getArgument(0).(Assignment).getRightOperand() = val.asExpr().getExpr()
```
would additionally match such methods, though we wouldn't automatically recognize subsequent calls to `foo.name` as "accesses" to that value, as far as I know.